ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ProposersDataManager.java

```diff
@@ -55,25 +64,56 @@ public PayloadAttributesCalculator(
     this.proposerDefaultFeeRecipient = proposerDefaultFeeRecipient;
   }
 
-  public void updateProposers(
-      final Collection<BeaconPreparableProposer> proposers, final UInt64 currentSlot) {
+  public void updatePreparedProposers(
+      final Collection<BeaconPreparableProposer> preparedProposers, final UInt64 currentSlot) {
     // Remove expired validators
-    proposerInfoByValidatorIndex.values().removeIf(info -> info.hasExpired(currentSlot));
+    final Iterator<Entry<UInt64, PreparedProposerInfo>> iterator =
+        preparedProposerInfoByValidatorIndex.entrySet().iterator();
+    while (iterator.hasNext()) {
+      final Entry<UInt64, PreparedProposerInfo> entry = iterator.next();
+      if (entry.getValue().hasExpired(currentSlot)) {
+        validatorRegistrationInfoByValidatorIndex.remove(entry.getKey());
+        iterator.remove();
```

Comment:
This is an example of a potential race condition. It's possible the entry isn't in `validatorRegistrationInfoByValidatorIndex` yet but is added later and we then wouldn't prune it. 

If they don't always line up then we probably need to handle expiry of validator registrations separately. It does look like we add to the maps separately so probably do need to just always treat them completely separately.

Reply:
yeah. I rechecked how the registration api against the builder infra is supposed to be used. It should follow a regular poll, so at the end i'm going to keeping them separated, mirroring the api separation.